### DISCUSSION TITLE
Multiple environments and migrations – best practices?

### DISCUSSION THREAD

**Discussion by mstade** [supabase/supabase#542] [Questions] [answered]
Hi supabase community – pleased to meet y'all!

I'm currently working on a backend using Vercel, which is working out quite nicely and we're happy with the way Vercel sets up development, preview (staging), and production (live) environments for us to easily keep things neatly separated as we move along during the development lifecycle. We're particularly happy with how it integrates with GitHub so we can easily just push to specific branches and have those automatically build and deploy in given environments.

We're going with postgresql as our database of choice because – let's just be real – it's amazing. We've also been evaluating supabase for the past few weeks and though it's obviously early days we like what we're seeing, spinning up a new database is so easy it makes our heads spin, and the minimalistic UI is quite frankly a breath of fresh air – kudos!

We're now trying to marry our environmental setup in Vercel with the dev/preview/prod environments, and figure that this _should_ map quite nicely to just having different projects for each environment – so this is how we've set it up:

```
Our organization
┣━ Development
┣━ Preview
┗━ Production
```

This neatly separates things and let's us do catastrophically dumb stuff to for example the development environment, without breaking the others. With this, we have some questions:

1. Would you recommend this setup?
2. We plan to do migrations by simply running migration SQL scripts in the target environment – is this how you'd recommend doing things?
3. How do we script supabase settings, is there a management API we could use or is _everything_ in the DB so we could set these things through the aforementioned migration scripts?
4. Do you have any pointers on best practices setting up CI when developing against supabase?

Many thanks for any pointers you may be able to provide, and keep up all the great work!

**Comment by kennethacohen** [comment]
+1, would love to know how other folks have different environments set up.

It seems like one path is just duplicating a Project and having one per env, but open to other ideas!

**Comment by Vadorequest** [comment]
+1 I'm also looking for something very similar, synching changes is a real pain for me and I wondered if writing migrations is possible.

The problem is wider than simply DB migrations, everything that touches the app settings should be "migratable". At the moment, there doesn't seem to be much to help us:
- We can't diff two projects and know what are the differences (tables, columns, settings, etc)
- We can't clone/duplicate an existing project

This makes it very difficult to maintain a consistent multi-environments system.

At the very least, having a page that prints the whole DB configuration as text would allow to perform a text diff between 2 projects and see what are the differences to correct them manually. This doesn't sound too complicated, but isn't the best solution either, more like a workaround.

I'd like to know if this is something that is intended to be supported, and around when.

**Comment by shorn** [comment]
I am working on a demo project that uses a Java tool called Flyway to manage my Supabase schema: [Cabbage](https://github.com/kopi-cloud/cabbage).

It has a write-up of how I plan to [manage Supabase schemas across multiple environments](https://github.com/kopi-cloud/cabbage/blob/main/doc/environment-config.md) (`dev`, `ci`, `tst`, `prd`), deployed via Github actions and Netlify.

**Comment by aileksandar** [comment]
Something like schema diff in pgadmin would help a lot
https://www.pgadmin.org/docs/pgadmin4/development/schema_diff.html#schema-diff-ddl-comparison-panel

**Comment by vicovaro** [comment]
This is what they wrote on their pricing page
What if I need one project for development and one for production?
We are working on multi-environment projects. For now, you can create a project for your development backend and production backend. We give you four free projects as part of our free tier. This means you could have two development and two production servers as part of your free tier.
https://supabase.io/pricing

So I just created two different project to make sure I don't delete a user by accident.

**Reply by aditodkar** [reply]
But how are you keeping both the applications in sync? Just curious to know

**Reply by vicovaro** [reply]
Manually, because it's impossible to keep them in sync automatically using the web UI they provided.

**Reply by mstade** [reply]
Migrating from one database environment to another can be very tricky, and especially trying to come up with a general solution for how to do it. For example, we (as in our company) would almost never want to migrate actual _data_ from our preview environment(s) to production, only schemas and potentially transforming production data to fit the new schemas. We might want to do it differently going from production to preview or a local environment though, in that case we may well want to migrate both schemas and data from production, for example to reproduce a bug.

We do migrations manually. It's time consuming work, but it also gives us an understanding both for the importance of spending a little time up front thinking about the data model as well as an incentive to not introduce breaking database changes unnecessarily. Most of our database changes are additive, meaning they don't destroy or transform any data already existing in the database.

I would very much welcome some tooling to assist with at least some of the more generic cases of database migrations, but I'm a bit weary of solutions pretending to provide a full scale migration with no supervision necessary.

**Reply by aditodkar** [reply]
So the thing is we will have to be careful while doing changes to the supabase DB right? What if by mistake column is deleted or some modifications happen in the table how to do rollback? Do you take frequent backups of your DB in supabase?

**Reply by mstade** [reply]
Supabase actually automatically backs up your database daily, which is quite nice. It's also trivial to restore a backup. We don't make changes in the Supabase UI in the production environment, and I think the UI has sufficient safeguards to avoid accidental  structural changes to the database frankly. I suppose making accidental data changes is easier, though we haven't had any issues with that as far as I know. Maybe  It would be nice though if one could disable UI features for structural and data changes altogether in an environment, while still allowing it with SQL scripts.

We only use SQL scripts to do migrations. Shamefully we don't have a very structured process around this, mainly because we very rarely do changes to the database structure these days, so we employ different strategies at times. One thing we almost always do though is use the diff tool in pgadmin. It's quite good to determine changes, but I wouldn't use it's output right off the bat without having a good look through it first. Maybe it's flawless and I'm just paranoid, I'm no expert with pgadmin, but yeah your mileage may vary. 🤪

**Comment by logemann** [comment]
Thinking about the same for my project. I will evaluate the following approach:

- deployment with Netlify (different GH branches via Netlify Branch deploys - so only one Netlify Site)
- one project per environment in supabase. We will start with a simple prod/stage setup
- flyway migrations via GH Actions

lets see if this will cut it.

**Reply by mstade** [reply]
Hey @logemann, how did you fare with flyaway migrations via GH Actions? I'd be interested in learning more about this.

**Comment by jacob-ruiz** [comment]
Rather than separate projects per environment I wish my project could just have a global dropdown to switch between environments. And then some UI to sync schema changes from dev to preview to prod.

Here's a mockup of what I'm imagining. This seems like it would match supabase's overall philosophy of preferring UI and ease of use:

<img width="1509" alt="Screen Shot 2022-02-10 at 8 51 30 AM" src="https://user-images.githubusercontent.com/3422246/153372737-079cd92e-2a70-4f62-987b-82d2a3a2f3e4.png">

For now I'll be setting up separate projects per environment. But trying to manually keep these in sync seems likely to lead to frustration.

**Reply by Anan7Codes** [reply]
> Rather than separate projects per environment I wish my project could just have a global dropdown to switch between environments. And then some UI to sync schema changes from dev to preview to prod.
> 
> Here's a mockup of what I'm imagining. This seems like it would match supabase's overall philosophy of preferring UI and ease of use:
> 
> <img alt="Screen Shot 2022-02-10 at 8 51 30 AM" width="1509" src="https://user-images.githubusercontent.com/3422246/153372737-079cd92e-2a70-4f62-987b-82d2a3a2f3e4.png">
> 
> For now I'll be setting up separate projects per environment. But trying to manually keep these in sync seems likely to lead to frustration.

Waiting for this feature as well!!

**Reply by MildTomato** [reply]
We've been rolling out Database Branching recently, which looks a lot like your mockup suggestion @jacob-ruiz 

https://supabase.com/docs/guides/platform/branching

**Comment by elliottetzkorn** [comment]
Following this. Maintaining separate projects is not scalable.

**Reply by elliottetzkorn** [reply]
Automatically connecting projects by DB, or having dev and production on the same project have issues as well - but there should at least be a simple UI on the site to both see the DB schema diff, + clone DB's to keep dev and prod in sync.

**Comment by starascendin** [comment]
Supabase has such garbage UX.

How do you expect devs to build on this in production when devs have to spend valuable time manually diffing and spinning up diff dbs for each env?

It's suppose to make devs lives easier, instead it makes our lives more difficult by introducing things that we dont need to do

**Reply by CitrusFruits** [reply]
"garbage UX" is a little harsh, not to mention insensitive. Supabase doesn't have bad UX, it just hasn't solved the problem you want it to yet. As mentioned in [mstade's comment above](https://github.com/supabase/supabase/discussions/542#discussioncomment-1694012), migrating from one database environment to another is tricky. That all being said, I think helping facilitate this is an opportunity the Supabase team definitely can help solve sometime in the future and would be an extremely valuable feature.

For the meantime though, you can connect directly to the postgres databases in your projects and use migration tools just like you would anywhere else. Perhaps something like [migra](https://github.com/djrobstep/migra) would be helpful for your use case.

**Reply by mstade** [reply]
I agree with @CitrusFruits – no need for such a harsh and unproductive tone. @kiwicopple and gang are doing solid work, and I'm sure they'll come up with some amazing solutions to common migration patterns as well. Rome wasn't built in a day, and all that.

**Reply by shorn** [reply]
> For the meantime though, you can connect directly to the postgres databases in your projects and use migration tools **just like you would anywhere else**. 

Definitely - this is how any experienced team will do it when environment and configuration management are serious concerns.

It's their money to spend how they want - but I worry that SB is burning too many resources on the UI re-building generic database tools that are handled better by dedicated tooling. On the other hand; a slick UI is a great pipeline for Supabase to encourage less experienced users who would otherwise bounce with no built-in tooling.  Catch-22, I guess.

**Comment by Morzaram** [comment]
Following. I feel like this topic is one of the only things holding me back from using them. Love everything else so far! Maybe it'll be addressed on launch week?

**Reply by MildTomato** [reply]
@Morzaram We've been tackling issue with Supabase Branching
check it out - https://supabase.com/docs/guides/platform/branching

**Comment by silentworks** [comment, marked as answer]
I think as @mstade has stated multiple environments and migrations is tricky to solve with a one size fits all solution. Personally, I go the way of using a migration library like [ley](https://github.com/lukeed/ley) to handle my database schema changes, row-level security (RLS) policies and database functions. This way when I spin up a new Supabase project (new environment), I just run the migration against it as you are given direct access to the database. The only part that currently is not easy to do via a script or anything is the settings in the dashboard for Auth and so forth. Hopefully, this will be addressed in the future and it will become even easier to get things fully automated. For my development database and staging, I run my migrations manually and for my production database, I set it to run from my CI tool (GitHub actions in this case). You can take a look at my project with all its code on GitHub https://github.com/silentworks/waiting-list.

I would also like to add that most database hosting platforms don't sync schema between databases for you as this is for the developers working on the project to control how they do so. ~Having Supabase do this for you could be more of a problem than a solution.~

1. Would you recommend this setup?
This setup looks fine to me, however, know that currently you are limited to 2 free projects per organisation but you can use the supabase cli to setup a local instance and use that as a development environment.

2. We plan to do migrations by simply running migration SQL scripts in the target environment – is this how you'd recommend doing things?
This is how I currently do it and how others I know are doing it too.

3. How do we script supabase settings, is there a management API we could use or is everything in the DB so we could set these things through the aforementioned migration scripts?
At the moment no.

4. Do you have any pointers on best practices setting up CI when developing against supabase?
Just keep your database password/secret key outside of your codebase and use environment variables with your CI system.

**Comment by steffenstolze** [comment]
Anything new on this topic since March?

**Reply by MildTomato** [reply]
We've been tackling issue with Supabase Branching
check it out - https://supabase.com/docs/guides/platform/branching

**Comment by ilbertt** [comment]
This topic is really interesting and, since I recently found out this powerful _Infrastructure as Code_ tool called [Pulumi](https://www.pulumi.com/), I'd suggest to investigate it further and try to implement a Supabase plugin for it. I think that, in order for Pulumi to work with Supabase, there should be some kind of _management API_ from Supabase, as @mstade suggested.
Looking forward to see the evolution of this topic!

**Comment by sampl** [comment]
+1, this is something Firebase does pretty well with its [Security Rules](https://firebase.google.com/docs/rules/manage-deploy), a config file in the project root. Just edit rules, deploy, and the db stays in sync.

> For the meantime though, you can connect directly to the postgres databases in your projects and use migration tools just like you would anywhere else.

> Personally, I go the way of using a migration library like [ley](https://github.com/lukeed/ley) to handle my database schema changes

It makes total sense to me that an existing tool might do this very well. IMO one of the benefits of Supabase is that the team selects a good external library and abstracts it for us (so I don’t have to pick and shoe horn it in). So, maybe its a matter of getting consensus and folding it into the official docs?

**Reply by SebasScript** [reply]
The supabase team made a tool based on pg admin schema diff to handle migrations between dev and prod, and also compared different tools. I read through this and was researching on different tools, pros and cons and came across their blog post on this

[https://supabase.com/blog/supabase-cli#migrations](url)

**Reply by sampl** [reply]
Oh interesting, thanks for sharing.

The warning at the top does say it’s a bit outdated, though it looks there’s some more relevant CLI docs here related to pushing db updates from local edits: https://supabase.com/docs/guides/local-development

**Reply by Anan7Codes** [reply]
I wish there was a command like supabase db pull to pull an existing project locally or should I use supabase migration commands to my locally running supabase project, make the changes and push it to my project in the cloud? Any idea?

**Reply by sampl** [reply]
There are new features in the CLI as of today, maybe this helps? https://supabase.com/blog/supabase-cli-v1-and-admin-api-beta

**Reply by Anan7Codes** [reply]
I did watch the video on YT but I still don't think there is a way to pull project locally and then make changes and push it back.

**Reply by abhay187** [reply]
@Anan7Codes yeah I am looking for something like this. Right now there seems to be no way to start using CLI once you manually created tables or made changes to your remote database.

Must needed feature.

**Reply by Anan7Codes** [reply]
Well, they've recently teamed up with Snaplet, and also Tyler replied to one of my comment on YT that this feature is coming soon so stay put.

**Reply by abhay187** [reply]
Thanks for the information. Snaplet seems promising. Will definitely give it a try.

**Comment by abhay187** [comment]
After reading this I realize how easy it is to work with NoSQL Database. No migration shit to worry about. I miss firebase and MongoDB a lot  😭.

**Reply by Anan7Codes** [reply]
It's true, there are obv pros and cons but Supabase is much more than just a DB which edge functions and vaults. They'll definitely sort this issue out cause it is one of the most requested feature.

**Reply by abhay187** [reply]
Let's hope they come up with something easier than the current approach. Right now the biggest issue is that if you already have made changes to Db through the dashboard you can't pull changes to local. 
As a developer, I myself can understand the circumstances and hope that the team will come up with something amazing.

**Reply by Anan7Codes** [reply]
I get you perfectly. I'm on a similar boat as well.

**Reply by MildTomato** [reply]
You can now pull changes from a remote project to your local env with this command
https://supabase.com/docs/reference/cli/supabase-db-pull

**Comment by MrChrisRodriguez** [comment]
Maybe I'm missing the mark, but hasn't PlanetScale solved this problem in a way that Supabase could learn from?

- [PlanetScale Workflow](https://planetscale.com/docs/concepts/planetscale-workflow)
  - [Branching](https://planetscale.com/docs/concepts/branching)
  - [Data Branching](https://planetscale.com/docs/concepts/data-branching)

**Reply by abhay187** [reply]
Planetscale is able to achieve this because they remove the Foreign key constraints.
https://planetscale.com/docs/learn/operating-without-foreign-key-constraints

Personally, I liked this approach coming from nosql background.

**Reply by MildTomato** [reply]
> but hasn't PlanetScale solved this problem in a way that Supabase could learn from?

We did learn from it 😄 

Some differences between Planetscale and Supabase Branching.

- Planetscale has a unique "dbuser" with it's own password for each Branch. The dbuser and password are what determines which Branch you connect to. It also supports multiple user/passwords per branch I think.
- Supabase has an isolated instance for each Preview branch with it's own DB URL, DB password, and set of API keys. The Preview branch behaves in the same way as your normal project (which is your prod branch).

There are probably good pros and cons to each approach, but I'll leave it for others to comment on it.

We haven't done any data branching yet, but schema branching is an early access feature today.

Check it out - https://supabase.com/docs/guides/platform/branching

**Comment by sampl** [comment]
Also might be totally unrelated, but it looks like prisma generates migrations from a schema config https://www.prisma.io/docs/concepts/components/prisma-schema

**Comment by jeremyisatrecharm** [comment]
I'd really like this as well. The CLI / running locally is an OK solution but the Docker container bogs down my machine so I'd prefer to just use another Supabase instance.

**Reply by jeremyisatrecharm** [reply]
Update: I upgraded to an M1 (now M2) mac and the Docker container runs fine (the M* macs are magical). Still, I find this to be a useful feature - at least because other team members still have Intel macs.

**Reply by ghost** [reply]
Runs great for me on a Ryzen 5600x with Manjaro Linux. The containers do use quite a bit of RAM though. If you're hitting your RAM limit, it will slow any machine down.

**Reply by jeremyisatrecharm** [reply]
Good point - my M2 has 32GB of RAM (I also briefly had an M1 with 32GB of RAM but then the M2 came out so I exchanged). My old Intel mac had 16GB and was almost unusable running the Supabase Docker :-(

**Reply by abhay187** [reply]
Apart from docker persistence is also a big pain. You have to keep it running even when you don't need it.

**Comment by ghost** [comment]
There is now a [Management API](https://supabase.com/docs/reference/api/introduction) in beta.

**Reply by abhay187** [reply]
Hey @waynebloss can you explain how this management API will improve local development and migration.

**Reply by ghost** [reply]
@abhay187 Oh, I thought I had the wrong discussion for a moment but the title of this one is  "Multiple environments and migrations".

So with regards to this topic, I would say that the relevance of my comment lies with the fact that the management API can be used to spin up and manage another environment....

**Comment by abhay187** [comment]
Does anyone realize that the migration schema generated by supabase is unreadable? I am wondering what if someone pushes a malicious code. I don't think anyone will catch this out. This is a serious security and privacy risk.

**Reply by bmadone** [reply]
Any updates? How you guys are handling multiple dev environements?

**Reply by abhay187** [reply]
Right now I am in the development phase and use hosted version. I also want to let you know that this is also my concern. I hope someone from the supabase team can answer this question.

**Reply by sweatybridge** [reply]
To clarify on the schema generation process, there are mainly two ways to generate migration schemas off a remote project
1. supabase db dump
2. supabase db diff

The first time `db remote commit` is run, it is actually running `db dump`. Then subsequently it uses `db diff`.

The difference is db dump uses `pg_dump` under the hood with a lot of flags to ignore entities in supabase managed schemas, including extensions, auth, storage, etc. There are also entities which pg_dump don't support ignoring. In those cases, we comment out those statements via `sed`. Typically on a small project, about 70% of the dumped lines are just comments. Do you think this is obscuring some of the actual migration statements?

When using `db diff`, the CLI runs `migra` under the hood. The generated schema is usually very concise because it does a schema diff instead of a full dump. The caveat is not all entities may be captured in the diff output so we usually recommend a manual review after.

If you find any specific sql statements that's too verbose to understand, feel free to post as a discussion after sanitizing and we can help decipher it.
